File: src/main/java/ua/mezik/socketchat/model/ChatMessage.java
package ua.mezik.socketchat.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.Data;

import java.sql.Timestamp;
import java.util.Calendar;

@Entity
@Data
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
public abstract class ChatMessage {
    @Id
    @GeneratedValue
    private long id;
    @ManyToOne
    private Account owner;
    @ManyToOne
    @JsonIgnore
    private Chat chat;
    private MessageType type;
    private Timestamp timestamp;

    public ChatMessage(Account owner, Chat chat, MessageType type) {
        this.owner = owner;
        this.chat = chat;
        this.type = type;
        this.timestamp = new Timestamp(Calendar.getInstance().getTime().getTime());
    }

    public ChatMessage() {
    }
}

File: src/main/kotlin/com/mezik/uchat/model/database/ChatMessage.kt
package com.mezik.uchat.model.database

import com.fasterxml.jackson.annotation.JsonIgnore
import com.mezik.uchat.model.message.MessageRequest
import com.mezik.uchat.service.SequencedIdGenerator
import org.bson.types.ObjectId
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.util.*

@Document("chat_message")
sealed class ChatMessage(
    val owner: Account,
    @field:JsonIgnore val chat: Chat,
    val type: MessageType,
    val timestamp: Long,
    val keyId: Long,
    @field:JsonIgnore @Id var databaseId: ObjectId = ObjectId(),
    var id: Long = SequencedIdGenerator.instance.generateId("chat_message")
) {
    companion object {
        fun createBasedOnRequest(request: MessageRequest, account: Account, chat: Chat): ChatMessage {
            val timestamp = Calendar.getInstance().timeInMillis
            return when (request.messageType) {
                MessageType.TEXT -> TextMessage(
                    account,
                    chat,
                    request.messageType,
                    timestamp,
                    request.keyId,
                    request.buffer,
                )

                MessageType.FILE -> FileMessage(
                    account,
                    chat,
                    request.messageType,
                    timestamp,
                    request.keyId,
                    request.buffer,
                )
            }
        }
    }
}
